model/PermissionsService.php

```diff
@@ -54,61 +52,148 @@ public function __construct(
         $this->eventManager = $eventManager;
     }
 
-    public function saveResourcePermissionsRecursive(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet
-    ): void {
-        $this->saveResourcePermissions($resource, $privilegesToSet, true);
-    }
-
-    private function saveResourcePermissions(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet,
-        bool $isRecursive
-    ): void {
-        $currentPrivileges = $this->dataBaseAccess->getResourcePermissions($resource->getUri());
-        $addRemove = $this->strategy->normalizeRequest($currentPrivileges, $privilegesToSet);
+    public function applyPermissions(ChangePermissionsCommand $command): void
+    {
+        $resources = $this->getResourcesToUpdate($command);
+        $currentPermissions = $this->getResourcesPermissions($resources);
+        $permissionsDelta = $this->getDeltaForResourceTree($command, $currentPermissions);
 
-        if (empty($addRemove)) {
+        if (empty($permissionsDelta)) {
             return;
         }
 
-        $resourcesToUpdate = $this->getResourcesToUpdate($resource, $isRecursive);
-        $permissionsList = $this->getResourcesPermissions($resourcesToUpdate);
-        $actions = $this->getActions($resourcesToUpdate, $permissionsList, $addRemove);
-
-        $this->dryRun($actions, $permissionsList);
+        $actions = $this->getActions($resources, $currentPermissions, $permissionsDelta);
+        $this->dryRun($actions, $currentPermissions);
         $this->wetRun($actions);
-        $this->triggerEvents($addRemove, $resource->getUri(), $isRecursive);
+
+        $this->triggerEvents(
+            $permissionsDelta[$command->getRoot()->getUri()],
+            $command->getRoot()->getUri(),
+            $command->isRecursive(),
+            $command->applyToNestedResources()
+        );
+    }
+
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
+    public function saveResourcePermissionsRecursive(
+        core_kernel_classes_Resource $resource,
+        array $privilegesToSet
+    ): void {
+        $command = new ChangePermissionsCommand($resource, $privilegesToSet);
+        $command->withRecursion();
+        $command->withNestedResources();
```

Comment:
question: why are we using setters instead of constructor? :thinking:

Reply:
Mostly to prevent passing flag parameters to the constructor.

Previous discussions about this:

- https://github.com/oat-sa/extension-tao-dac-simple/pull/211#discussion_r1132229645
- https://github.com/oat-sa/extension-tao-dac-simple/pull/211#discussion_r1132076840

It can be argued that using setters like these make the code slightly easier to read because removes the usage of flag parameters in the constructor (i.e. passing bools like `new ChangePermissionsCommand($resource, $privilegesToSet, true, true)` would make  the code a bit less readable). However I don't strongly prefer one way vs another, and even the former solution used flag parameters for the command constructor ([here](https://github.com/oat-sa/extension-tao-dac-simple/pull/210/commits/8abe05e0917e51da81b17faecd77873346446a8b)).

Reply:
I do prefer the constructor too, but since it can be "opiniated", I think we can let as is and move forward. All good? :)